src/Temporalio/Bridge/src/worker.rs

```diff
@@ -0,0 +1,402 @@
+use crate::client::Client;
+use crate::runtime::Runtime;
+use crate::ByteArray;
+use crate::ByteArrayRef;
+use crate::UserDataHandle;
+use prost::Message;
+use temporal_sdk_core::WorkerConfigBuilder;
+use temporal_sdk_core_api::errors::PollActivityError;
+use temporal_sdk_core_api::errors::PollWfError;
+use temporal_sdk_core_api::Worker as CoreWorker;
+use temporal_sdk_core_protos::coresdk::workflow_completion::WorkflowActivationCompletion;
+use temporal_sdk_core_protos::coresdk::ActivityHeartbeat;
+use temporal_sdk_core_protos::coresdk::ActivityTaskCompletion;
+
+use std::sync::Arc;
+use std::time::Duration;
+
+#[repr(C)]
+pub struct WorkerOptions {
+    namespace: ByteArrayRef,
+    task_queue: ByteArrayRef,
+    build_id: ByteArrayRef,
+    identity_override: ByteArrayRef,
+    max_cached_workflows: u32,
+    max_outstanding_workflow_tasks: u32,
+    max_outstanding_activities: u32,
+    max_outstanding_local_activities: u32,
+    max_concurrent_workflow_task_polls: u32,
+    nonsticky_to_sticky_poll_ratio: f32,
+    max_concurrent_activity_task_polls: u32,
+    no_remote_activities: bool,
+    sticky_queue_schedule_to_start_timeout_millis: u64,
+    max_heartbeat_throttle_interval_millis: u64,
+    default_heartbeat_throttle_interval_millis: u64,
+    max_activities_per_second: f64,
+    max_task_queue_activities_per_second: f64,
+}
+
+#[derive(Clone)]
+pub struct Worker {
+    worker: Option<Arc<temporal_sdk_core::Worker>>,
+    runtime: Runtime,
+}
+
+/// Only runtime or fail will be non-null. Whichever is must be freed when done.
+#[repr(C)]
+pub struct WorkerOrFail {
+    worker: *mut Worker,
+    fail: *const ByteArray,
+}
+
+macro_rules! enter_sync {
+    ($runtime:expr) => {
+        temporal_sdk_core::telemetry::set_trace_subscriber_for_current_thread(
+            $runtime.core.trace_subscriber(),
+        );
+        let _guard = $runtime.core.tokio_handle().enter();
+    };
+}
+
+#[no_mangle]
+pub extern "C" fn worker_new(client: *mut Client, options: *const WorkerOptions) -> WorkerOrFail {
+    let client = unsafe { &mut *client };
+    enter_sync!(client.runtime);
+    let options = unsafe { &*options };
+
+    let (worker, fail) = match options.try_into() {
+        Err(err) => (
+            std::ptr::null_mut(),
+            client
+                .runtime
+                .alloc_utf8(&format!("Invalid options: {}", err))
+                .into_raw()
+                .cast_const(),
+        ),
+        Ok(config) => match temporal_sdk_core::init_worker(
+            &client.runtime.core,
+            config,
+            client.core.clone().into_inner(),
+        ) {
+            Err(err) => (
+                std::ptr::null_mut(),
+                client
+                    .runtime
+                    .alloc_utf8(&format!("Worker start failed: {}", err))
```

Comment:
I'd throw the proper exception type here.
Generally you're going to need a mechanism to pass around the error types from Rust to C#, it's something worth doing.

Reply:
Haven't needed this on Python. What case does the user need specific exceptions here? Invalid operation covers all cases until there is a compelling reason for a more specific exception type for a more specific situation.

Reply:
I just noticed I never replied to this.
My ask here was that if worker init fails for example because it failed to connect, this would result in something other than invalid operation because it's not.

Reply:
In .NET invalid operation is general purpose. Also, workflow construction is synchronous, so is it really connecting? We use the same general purpose exception to represent all workflow construction failures in Python and I figure we should here too unless there's a good reason not to.

Reply:
You're good, I double checked. I think this used to connect which is what had me confused.